README.md
# useWhisper

React Hook for OpenAI Whisper API with speech recorder, real-time transcription and silence removal built-in

---

- ### Demo

- ###### Real-Time transcription demo

https://user-images.githubusercontent.com/2707253/224465747-0b1ee159-21dd-4cd0-af9d-6fc9b882d716.mp4

---

- ### Announcement

  useWhisper for React Native is being developed.

Repository: [https://github.com/ginixsan/use-whisper-native](https://github.com/ginixsan/use-whisper-native)

Progress: [https://github.com/ginixsan/use-whisper-native/issues/1](https://github.com/ginixnsan/use-whisper-native/issues/1)

- ### Install

```
npm i @ginixsan/use-whisper
```

```
yarn add @ginixsan/use-whisper
```

- ### Usage

```jsx
import { useWhisper } from '@ginixsan/use-whisper'

const App = () => {
  const {
    recording,
    speaking,
    transcribing,
    transcript,
    pauseRecording,
    startRecording,
    stopRecording,
  } = useWhisper({
    apiKey: process.env.OPENAI_API_TOKEN, // YOUR_OPEN_AI_TOKEN
  })

  return (
    <div>
      <p>Recording: {recording}</p>
      <p>Speaking: {speaking}</p>
      <p>Transcribing: {transcribing}</p>
      <p>Transcribed Text: {transcript.text}</p>
      <button onClick={() => startRecording()}>Start</button>
      <button onClick={() => pauseRecording()}>Pause</button>
      <button onClick={() => stopRecording()}>Stop</button>
    </div>
  )
}
```

- ###### Custom Server (keep OpenAI API token secure)

```jsx
import { useWhisper } from '@ginixsan/use-whisper'

const App = () => {
  /**
   * you have more control like this
   * do whatever you want with the recorded speech
   * send it to your own custom server
   * and return the response back to useWhisper
   */
  const onTranscribe = (blob: Blob) => {
    const base64 = await new Promise<string | ArrayBuffer | null>(
      (resolve) => {
        const reader = new FileReader()
        reader.onloadend = () => resolve(reader.result)
        reader.readAsDataURL(blob)
      }
    )
    const body = JSON.stringify({ file: base64, model: 'whisper-1' })
    const headers = { 'Content-Type': 'application/json' }
    const { default: axios } = await import('axios')
    const response = await axios.post('/api/whisper', body, {
      headers,
    })
    const { text } = await response.data
    // you must return result from your server in Transcript format
    return {
      blob,
      text,
    }
  }

  const { transcript } = useWhisper({
    // callback to handle transcription with custom server
    onTranscribe,
  })

  return (
    <div>
      <p>{transcript.text}</p>
    </div>
  )
}
```

- ### Examples

- ###### Real-time streaming trascription

```jsx
import { useWhisper } from '@ginixsan/use-whisper'

const App = () => {
  const { transcript } = useWhisper({
    apiKey: process.env.OPENAI_API_TOKEN, // YOUR_OPEN_AI_TOKEN
    streaming: true,
    timeSlice: 1_000, // 1 second
    whisperConfig: {
      language: 'en',
    },
  })

  return (
    <div>
      <p>{transcript.text}</p>
    </div>
  )
}
```

- ###### Remove silence before sending to Whisper to save cost

```jsx
import { useWhisper } from '@ginixsan/use-whisper'

const App = () => {
  const { transcript } = useWhisper({
    apiKey: process.env.OPENAI_API_TOKEN, // YOUR_OPEN_AI_TOKEN
    // use ffmpeg-wasp to remove silence from recorded speech
    removeSilence: true,
  })

  return (
    <div>
      <p>{transcript.text}</p>
    </div>
  )
}
```

- ###### Auto start recording on component mounted

```jsx
import { useWhisper } from '@ginixsan/use-whisper'

const App = () => {
  const { transcript } = useWhisper({
    apiKey: process.env.OPENAI_API_TOKEN, // YOUR_OPEN_AI_TOKEN
    // will auto start recording speech upon component mounted
    autoStart: true,
  })

  return (
    <div>
      <p>{transcript.text}</p>
    </div>
  )
}
```

- ###### Keep recording as long as the user is speaking

```jsx
import { useWhisper } from '@ginixsan/use-whisper'

const App = () => {
  const { transcript } = useWhisper({
    apiKey: process.env.OPENAI_API_TOKEN, // YOUR_OPEN_AI_TOKEN
    nonStop: true, // keep recording as long as the user is speaking
    stopTimeout: 5000, // auto stop after 5 seconds
  })

  return (
    <div>
      <p>{transcript.text}</p>
    </div>
  )
}
```

- ###### Customize Whisper API config when autoTranscribe is true

```jsx
import { useWhisper } from '@ginixsan/use-whisper'

const App = () => {
  const { transcript } = useWhisper({
    apiKey: process.env.OPENAI_API_TOKEN, // YOUR_OPEN_AI_TOKEN
    autoTranscribe: true,
    whisperConfig: {
      prompt: 'previous conversation', // you can pass previous conversation for context
      response_format: 'text', // output text instead of json
      temperature: 0.8, // random output
      language: 'es', // Spanish
    },
  })

  return (
    <div>
      <p>{transcript.text}</p>
    </div>
  )
}
```

- ### Dependencies

  - **@ginixsan/react-hooks-async** asynchronous react hooks
  - **recordrtc:** cross-browser audio recorder
  - **lamejs** encode wav into mp3 for cross-browser support
  - **@ffmpeg/ffmpeg:** for silence removal feature
  - **hark:** for speaking detection
  - **axios:** since fetch does not work with Whisper endpoint

_most of these dependecies are lazy loaded, so it is only imported when it is needed_

- ### API

- ###### Config Object

| Name            | Type                                               | Default Value  | Description                                                                                                          |
| --------------- | -------------------------------------------------- | -------------- | -------------------------------------------------------------------------------------------------------------------- |
| apiKey          | string                                             | ''             | your OpenAI API token                                                                                                |
| autoStart       | boolean                                            | false          | auto start speech recording on component mount                                                                       |
| autoTranscribe  | boolean                                            | true           | should auto transcribe after stop recording                                                                          |
| mode            | string                                             | transcriptions | control Whisper mode either transcriptions or translations, currently only support translation to English            |
| nonStop         | boolean                                            | false          | if true, record will auto stop after stopTimeout. However if user keep on speaking, the recorder will keep recording |
| removeSilence   | boolean                                            | false          | remove silence before sending file to OpenAI API                                                                     |
| stopTimeout     | number                                             | 5,000 ms       | if nonStop is true, this become required. This control when the recorder auto stop                                   |
| streaming       | boolean                                            | false          | transcribe speech in real-time based on timeSlice                                                                    |
| timeSlice       | number                                             | 1000 ms        | interval between each onDataAvailable event                                                                          |
| whisperConfig   | [WhisperApiConfig](#whisperapiconfig)              | undefined      | Whisper API transcription config                                                                                     |
| onDataAvailable | (blob: Blob) => void                               | undefined      | callback function for getting recorded blob in interval between timeSlice                                            |
| onTranscribe    | (blob: Blob) => Promise<[Transcript](#transcript)> | undefined      | callback function to handle transcription on your own custom server                                                  |

- ###### WhisperApiConfig

| Name            | Type   | Default Value | Description                                                                                                                                                                                                                                                                                                                                               |
| --------------- | ------ | ------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| prompt          | string | undefined     | An optional text to guide the model's style or continue a previous audio segment. The prompt should match the audio language.                                                                                                                                                                                                                             |
| response_format | string | json          | The format of the transcript output, in one of these options: json, text, srt, verbose_json, or vtt.                                                                                                                                                                                                                                                      |
| temperature     | number | 0             | The sampling temperature, between 0 and 1. Higher values like 0.8 will make the output more random, while lower values like 0.2 will make it more focused and deterministic. If set to 0, the model will use [log probability](https://en.wikipedia.org/wiki/Log_probability) to automatically increase the temperature until certain thresholds are hit. |
| language        | string | en            | The language of the input audio. Supplying the input language in [ISO-639-1](https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes) format will improve accuracy and latency.                                                                                                                                                                             |

- ###### Return Object

| Name           | Type                      | Description                                                               |
| -------------- | ------------------------- | ------------------------------------------------------------------------- |
| recording      | boolean                   | speech recording state                                                    |
| speaking       | boolean                   | detect when user is speaking                                              |
| transcribing   | boolean                   | while removing silence from speech and send request to OpenAI Whisper API |
| transcript     | [Transcript](#transcript) | object return after Whisper transcription complete                        |
| pauseRecording | Promise                   | pause speech recording                                                    |
| startRecording | Promise                   | start speech recording                                                    |
| stopRecording  | Promise                   | stop speech recording                                                     |

- ###### Transcript

| Name | Type   | Description                                |
| ---- | ------ | ------------------------------------------ |
| blob | Blob   | recorded speech in JavaScript Blob         |
| text | string | transcribed text returned from Whisper API |

- ### Roadmap

  - react-native support, will be export as use-whisper/native

---

**_Contact me for web or mobile app development using React or React Native_**  
[https://ginixsan.github.io](https://ginixsan.github.io)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.2.0] - 2023-04-01

### Added

- react-native support
- demo web app
- server app example
- support more audio output format
- able to run custom ffmpeg command
- expose onTranscribing event

## [0.2.0] - 2023-03-18

### Added

- lamejs as new dependency

### Changed

- audio recorded from microphone will be in lossless .wav format and will be converted to mp3 for cross-browser suppport
- recorded mp3 be optimized for speech with bitrate of 96k and sample rate of 44.1k

### Fixed

- audio will be in mp3 format across all major browsers like Chrome, Safari, Firefox, Edge. [#13](https://github.com/ginixsan/use-whisper/issues/13) [#10](https://github.com/ginixsan/use-whisper/issues/10)

## [0.1.2] - 2023-03-15

### Added

- add missing mode to onWhispered dependency list

### Changed

- remove excessive use of useCallback

## [0.1.1] - 2023-03-14

### Added

- add mode option for Whisper API, choose either transcriptions or translations
  (currently only support translation to English)

### Changed

- default timeSlice from 2000ms to 1000ms

## [0.1.0] - 2023-03-11

### Added

- streaming option for real-time trascription
- timeSlice option to control onDataAvailable event
- onDataAvaiable option for getting recorded blob in interval based on timeSlice

### Changed

- recording in higher audio quality to help Whisper in transcription

### Removed

- customServer option, deprecated since 0.0.11

## [0.0.12] - 2023-03-09

### Changed

- autoTranscribe default to true
- update examples in README.md

## [0.0.11] - 2023-03-08

### Added

- add onTranscribe callback for handling transcription using custom server
- add whisperConfig to control Whisper API when trascribing

### Changed

- autoTranscribe config, default to false. useWhisper will not longer try to transcribe after recorder stopped.

### Removed

- deprecated customServer, use onTranscribe instead

## [0.0.10] - 2023-03-08

### Added

- this changelog file
- comments to every functions and variables
- @ginixsan/react-hooks-async for useCallbackAsync and useEffectAsync
- add Github Actions CI/CD

### Fixed

- now will not send audio to Whisper if converted mp3 audio is empty

### Changed

- transcripting state changed to transcribing to make it a valid word
- move constants string to configs.ts

### Removed

- hooks directory, now use @ginixsan/react-hooks-async package instead
- remove console.log from distribution build


src/configs.js
/**
 * default useWhisper configuration
 */

/**
 * default timeout for recorder
 */
const defaultTimeout = {
  stop: 5_000, // Default stop timeout, update as needed
}

/**
 * default transcript object
 */
const defaultTranscript = {
  blob: null, // Default value
  text: '', // Default value
}

export const defaultStopTimeout = 5_000

export const ffmpegCoreUrl =
  'https://unpkg.com/@ffmpeg/core-st@0.11.1/dist/ffmpeg-core.js'

export const silenceRemoveCommand =
  'silenceremove=start_periods=1:stop_periods=-1:start_threshold=-30dB:stop_threshold=-30dB:start_silence=2:stop_silence=2'

export const whisperApiEndpoint = 'https://api.openai.com/v1/audio/'
export const defaultConfig = {
  apiKey: null, // Default value
  autoStart: false, // Assuming default is false, update as needed
  autoTranscribe: false, // Assuming default is false, update as needed
  mode: 'defaultMode', // Replace with actual default mode
  nonStop: false, // Assuming default is false, update as needed
  removeSilence: false, // Assuming default is false, update as needed
  stopTimeout: defaultStopTimeout, // Default value from configs
  streaming: false, // Assuming default is false, update as needed
  timeSlice: 1000, // Default timeslice, update as needed
  onDataAvailable: null, // Default value
  onTranscribe: null, // Default value
}


src/useWhisper.js
import { useEffectAsync, useMemoAsync } from '@chengsokdara/react-hooks-async'

import { useEffect, useRef, useState } from 'react'

import {
  defaultStopTimeout,
  ffmpegCoreUrl,
  silenceRemoveCommand,
  whisperApiEndpoint,
} from './configs.js' // Correct relative path to configs.js

const defaultConfig = {
  apiKey: null,
  autoStart: false,
  autoTranscribe: false,
  mode: 'defaultMode',
  nonStop: false,
  removeSilence: false,
  stopTimeout: defaultStopTimeout,
  streaming: false,
  timeSlice: 1000,
  onDataAvailable: null,
  onTranscribe: null,
}

/**
 * default timeout for recorder
 */
const defaultTimeout = {
  stop: 5000, // Assuming 5000 milliseconds as default, update as needed
}

/**
 * default transcript object
 */
const defaultTranscript = {
  blob: null, // Assuming null as the default value
  text: '', // Assuming an empty string as the default value
}

/**
 * React Hook for OpenAI Whisper
 */
export const useWhisper = (config) => {
  const {
    apiKey,
    autoStart,
    autoTranscribe,
    mode,
    nonStop,
    removeSilence,
    stopTimeout,
    streaming,
    timeSlice,
    whisperConfig,
    onDataAvailable,
    onTranscribe,
  } = {
    ...defaultConfig,
    ...config,
  }

  if (!apiKey && !onTranscribeCallback) {
    throw new Error('apiKey is required if onTranscribe is not provided')
  }

  const chunks = useRef([])
  const encoder = useRef()
  const listener = useRef()
  const recorder = useRef()
  const stream = useRef()
  const timeout = useRef(defaultTimeout)

  const [recording, setRecording] = useState(false)
  const [speaking, setSpeaking] = useState(false)
  const [transcribing, setTranscribing] = useState(false)
  const [transcript, setTranscript] = useState(defaultTranscript)

  /**
   * cleanup on component unmounted
   * - flush out and cleanup lamejs encoder instance
   * - destroy recordrtc instance and clear it from ref
   * - clear setTimout for onStopRecording
   * - clean up hark speaking detection listeners and clear it from ref
   * - stop all user's media steaming track and remove it from ref
   */
  useEffect(() => {
    return () => {
      if (chunks.current) {
        chunks.current = []
      }
      if (encoder.current) {
        encoder.current.flush()
        encoder.current = undefined
      }
      if (recorder.current) {
        recorder.current.destroy()
        recorder.current = undefined
      }
      onStopTimeout('stop')
      if (listener.current) {
        // @ts-ignore
        listener.current.off('speaking', onStartSpeaking)
        // @ts-ignore
        listener.current.off('stopped_speaking', onStopSpeaking)
      }
      if (stream.current) {
        stream.current.getTracks().forEach((track) => track.stop())
        stream.current = undefined
      }
    }
  }, [])

  /**
   * if config.autoStart is true
   * start speech recording immediately upon component mounted
   */
  useEffectAsync(async () => {
    if (autoStart) {
      await onStartRecording()
    }
  }, [autoStart])

  /**
   * start speech recording and start listen for speaking event
   */
  const startRecording = async () => {
    await onStartRecording()
  }

  /**
   * pause speech recording also stop media stream
   */
  const pauseRecording = async () => {
    await onPauseRecording()
  }

  /**
   * stop speech recording and start the transcription
   */
  const stopRecording = async () => {
    await onStopRecording()
  }

  /**
   * start speech recording event
   * - first ask user for media stream
   * - create recordrtc instance and pass media stream to it
   * - create lamejs encoder instance
   * - check recorder state and start or resume recorder accordingly
   * - start timeout for stop timeout config
   * - update recording state to true
   */
  const onStartRecording = async () => {
    try {
      if (!stream.current) {
        await onStartStreaming()
      }
      if (stream.current) {
        if (!recorder.current) {
          const { RecordRTCPromisesHandler, StereoAudioRecorder } =
            await import('recordrtc')
          const recorderConfig = {
            mimeType: 'audio/webm', // Assuming a mimeType, update as needed
            numberOfAudioChannels: 1, // Assuming mono, update as needed
            recorderType: StereoAudioRecorder, // Assuming recorderType, update as needed
            sampleRate: 44100, // Assuming sampleRate = 44.1khz
            timeSlice: timeSlice || 1000, // Assuming a default timeSlice, update as needed
            type: 'audio', // Assuming type, update as needed
            ondataavailable: streaming ? onDataAvailable : undefined,
          }
          recorder.current = new RecordRTCPromisesHandler(
            stream.current,
            recorderConfig,
          )
        }
        if (!encoder.current) {
          const { Mp3Encoder } = await import('lamejs')
          encoder.current = new Mp3Encoder(1, 44100, 96) // mono, 44.1khz, 96kbps
        }
        const recordState = await recorder.current.getState()
        if (recordState === 'inactive' || recordState === 'stopped') {
          await recorder.current.startRecording()
        }
        if (recordState === 'paused') {
          await recorder.current.resumeRecording()
        }
        if (nonStop) {
          onStartTimeout('stop')
        }
        setRecording(true)
      }
    } catch (err) {
      console.error(err)
    }
  }

  /**
   * get user media stream event
   * - try to stop all previous media streams
   * - ask user for media stream with a system popup
   * - register hark speaking detection listeners
   */
  const onStartStreaming = async () => {
    try {
      // Stop any existing tracks
      if (stream.current) {
        stream.current.getTracks().forEach((track) => track.stop())
      }

      // Request access to the user's microphone
      stream.current = await navigator.mediaDevices.getUserMedia({
        audio: true, // Assuming 'audio' is a boolean or valid audio configuration object
      })

      // Initialize hark listener
      if (!listener.current) {
        const hark = await import('hark').then((m) => m.default) // Correcting the dynamic import of hark
        listener.current = hark(stream.current, {
          interval, // Ensure 'interval' is defined
          play, // Ensure 'play' is defined
        })
        listener.current.on('speaking', onStartSpeaking) // Ensure 'onStartSpeaking' is defined
        listener.current.on('stopped_speaking', onStopSpeaking) // Ensure 'onStopSpeaking' is defined
      }
    } catch (err) {
      console.error(err)
    }
  }

  /**
   * start stop timeout event
   */
  const onStartTimeout = (UseWhisperTimeout) => {
    if (!timeout.current[type]) {
      timeout.current[type] = setTimeout(onStopRecording, stopTimeout)
    }
  }

  /**
   * user start speaking event
   * - set speaking state to true
   * - clear stop timeout
   */
  const onStartSpeaking = () => {
    console.log('start speaking')
    setSpeaking(true)
    onStopTimeout('stop')
  }

  /**
   * user stop speaking event
   * - set speaking state to false
   * - start stop timeout back
   */
  const onStopSpeaking = () => {
    console.log('stop speaking')
    setSpeaking(false)
    if (nonStop) {
      onStartTimeout('stop')
    }
  }

  /**
   * pause speech recording event
   * - if recorder state is recording, pause the recorder
   * - clear stop timeout
   * - set recoriding state to false
   */
  const onPauseRecording = async () => {
    try {
      if (recorder.current) {
        const recordState = await recorder.current.getState()
        if (recordState === 'recording') {
          await recorder.current.pauseRecording()
        }
        onStopTimeout('stop')
        setRecording(false)
      }
    } catch (err) {
      console.error(err)
    }
  }

  /**
   * stop speech recording event
   * - flush out lamejs encoder and set it to undefined
   * - if recorder state is recording or paused, stop the recorder
   * - stop user media stream
   * - clear stop timeout
   * - set recording state to false
   * - start Whisper transcription event
   * - destroy recordrtc instance and clear it from ref
   */
  const onStopRecording = async () => {
    try {
      if (recorder.current) {
        const recordState = await recorder.current.getState()
        if (recordState === 'recording' || recordState === 'paused') {
          await recorder.current.stopRecording()
        }
        onStopStreaming()
        onStopTimeout('stop')
        setRecording(false)
        if (autoTranscribe) {
          await onTranscribing()
        } else {
          const blob = await recorder.current.getBlob()
          setTranscript({
            blob,
          })
        }
        await recorder.current.destroy()
        chunks.current = []
        if (encoder.current) {
          encoder.current.flush()
          encoder.current = undefined
        }
        recorder.current = undefined
      }
    } catch (err) {
      console.error(err)
    }
  }

  /**
   * stop media stream event
   * - remove hark speaking detection listeners
   * - stop all media stream tracks
   * - clear media stream from ref
   */
  const onStopStreaming = () => {
    if (listener.current) {
      // @ts-ignore
      listener.current.off('speaking', onStartSpeaking)
      // @ts-ignore
      listener.current.off('stopped_speaking', onStopSpeaking)
      listener.current = undefined
    }
    if (stream.current) {
      stream.current.getTracks().forEach((track) => track.stop())
      stream.current = undefined
    }
  }

  /**
   * stop timeout event
   * - clear stop timeout and remove it from ref
   */
  const onStopTimeout = (UseWhisperTimeout) => {
    if (timeout.current[type]) {
      clearTimeout(timeout.current[type])
      timeout.current[type] = undefined
    }
  }

  /**
   * start Whisper transcrition event
   * - make sure recorder state is stopped
   * - set transcribing state to true
   * - get audio blob from recordrtc
   * - if config.removeSilence is true, load ffmpeg-wasp and try to remove silence from speec
   * - if config.customServer is true, send audio data to custom server in base64 string
   * - if config.customServer is false, send audio data to Whisper api in multipart/form-data
   * - set transcript object with audio blob and transcription result from Whisper
   * - set transcribing state to false
   */
  const onTranscribing = async () => {
    console.log('transcribing speech')
    try {
      if (encoder.current && recorder.current) {
        const recordState = await recorder.current.getState()
        if (recordState === 'stopped') {
          setTranscribing(true)
          let blob = await recorder.current.getBlob()
          if (removeSilence) {
            const { createFFmpeg } = await import('@ffmpeg/ffmpeg')
            const ffmpeg = createFFmpeg({
              mainName: 'main',
              corePath: ffmpegCoreUrl,
              log: true,
            })
            if (!ffmpeg.isLoaded()) {
              await ffmpeg.load()
            }
            const buffer = await blob.arrayBuffer()
            console.log({ in: buffer.byteLength })
            ffmpeg.FS('writeFile', 'in.wav', new Uint8Array(buffer))
            await ffmpeg.run(
              '-i',
              'in.wav',
              '-acodec',
              'libmp3lame',
              '-b:a',
              '96k',
              '-ar',
              '44100',
              '-af',
              silenceRemoveCommand,
              'out.mp3',
            )
            const out = ffmpeg.FS('readFile', 'out.mp3')
            console.log({ out: out.buffer.byteLength })
            if (out.length <= 225) {
              ffmpeg.exit()
              setTranscript({ blob })
              setTranscribing(false)
              return
            }
            blob = new Blob([out.buffer], { type: 'audio/mpeg' })
            ffmpeg.exit()
          } else {
            const buffer = await blob.arrayBuffer()
            console.log({ wav: buffer.byteLength })
            const mp3 = encoder.current.encodeBuffer(new Int16Array(buffer))
            blob = new Blob([mp3], { type: 'audio/mpeg' })
            console.log({ blob, mp3: mp3.byteLength })
          }
          if (typeof onTranscribeCallback === 'function') {
            const transcribed = await onTranscribeCallback(blob)
            console.log('onTranscribe', transcribed)
            setTranscript(transcribed)
          } else {
            const file = new File([blob], 'speech.mp3', { type: 'audio/mpeg' })
            const text = await onWhispered(file)
            console.log('onTranscribing', { text })
            setTranscript({ blob, text })
          }
          setTranscribing(false)
        }
      }
    } catch (err) {
      console.info(err)
      setTranscribing(false)
    }
  }

  /**
   * Send audio file to Whisper to be transcribed
   * - create formdata and append file, model, and language
   * - append more Whisper config if whisperConfig is provided
   * - add OpenAPI Token to header Authorization Bearer
   * - post with axios to OpenAI Whisper transcript endpoint
   * - return transcribed text result
   */
  const onWhispered = useMemoAsync(
    async (file) => {
      // Whisper only accepts multipart/form-data currently
      const body = new FormData()
      body.append('file', file)
      body.append('model', 'whisper-1')
      if (mode === 'transcriptions') {
        body.append('language', whisperConfig?.language ?? 'en')
      }
      if (whisperConfig?.prompt) {
        body.append('prompt', whisperConfig.prompt)
      }
      if (whisperConfig?.response_format) {
        body.append('response_format', whisperConfig.response_format)
      }
      if (whisperConfig?.temperature) {
        body.append('temperature', `${whisperConfig.temperature}`)
      }
      const headers = {}
      headers['Content-Type'] = 'multipart/form-data'
      if (apiKey) {
        headers['Authorization'] = `Bearer ${apiKey}`
      }
      const axios = await import('axios').then((m) => m.default) // Correcting the dynamic import of axios
      const response = await axios.post(whisperApiEndpoint + mode, body, {
        headers,
      })
      return response.data.text
    },
    [apiKey, mode, whisperConfig],
  )

  return {
    recording,
    speaking,
    transcribing,
    transcript,
    pauseRecording,
    startRecording,
    stopRecording,
  }
}
